### PR TITLE
Requeue dead jobs properly

### LINKED ISSUE #135: Jobs in correct processing state are requeued after InvisibilityTimeout, which leads to stuck jobs after ungraceful shutdown
Hello,

First of all, thanks for developing and maintaining this project.

We've been long users of Hangfire with PostgreSQL. Recently, we've started evaluating the switch to Redis, and naturally, this lib was our first choice. After testing it for a while, I've found an edge case that can cause jobs to be lost by being stuck in the `Processing` state. Even though it's not common, it has a high impact.

How to reproduce it:
1. Set `InvisibilityTimeout` to `TimeSpan.FromSeconds(1)`
2. Enqueue job `_backgroundJobs.Enqueue(() => _testService.LongTask(TimeSpan.FromSeconds(600), CancellationToken.None));`
3. The job is started after a while, and we have `Task started` in the logs
4. `FetchedJobsWatcher` kicks in after a while with ` (Hangfire.Redis.StackExchange.FetchedJobsWatcher) 1 timed out jobs were found in the 'default' queue and re-queued.`
5. The job is started once again. `Task started` is shown once again in the logs
6. Kill the Hangfire server ungracefully with `kill -9 PID`
7. At this point, the job is stuck in the `Processing` state. It won't be requeued even after waiting for `InvisibilityTimeout`.

Additionally, I've noticed that at some point between points 5 and 6, Redis lists for a queue are absent, i.e., `keys *default*` returns nothing. That's why the job won't be picked up after starting a new server, even after the killed server is removed from the list of active servers.

Here is the job I've been using for testing:
```
    public async Task LongTask(TimeSpan duration, CancellationToken cancellationToken)
    {
        _logger.LogInformation("Task started");
        while (duration >= TimeSpan.Zero)
        {
            cancellationToken.ThrowIfCancellationRequested();

            var waitInterval = TimeSpan.FromSeconds(10);

            _logger.LogInformation("Waiting...");
            await Task.Delay(waitInterval, CancellationToken.None);
            duration -= duration;
            _logger.LogInformation("Wait finished");
        }
        _logger.LogInformation("Task finished");
    }
```

My questions are:
Is there a way to mitigate this?
Should `FetchedJobsWatcher` requeue the job when the job is in the `Processing` state with a live server executing it?
Is there some mechanism for monitoring `Processing` jobs for being stuck if it's okay for jobs to be removed from the queue?

Let me know if I can help by providing more testing and information.

### PR DESCRIPTION
**Pull request by Lexy2**
Fixes #135 

This PR includes several changes:
1. When a `RedisFetchedJob` is created / fetched, it also records the `Fetched` time not only in its `Fetched` hash entry, but also as a property in its object. As such, when the job is requeued or removed from queue, its `Fetched` hash entry is only removed if the values match (i.e. if there was no other process that updated this value before the current process). The same logic applies to the `dequeued` list.
- The two concurrent processes trying to remove and requeue the jobs are `ServerJobCancellationWatcher` and `FetchedJobsWatcher`.

2. When `FetchedJobsWatcher` checks whether the job should be requeued, it also checks if this job is being executed on a live server (i.e. the server that has not been removed from the list of servers by `ServerWatchdog`). **Note** that it will not requeue the job that is being executed on a possibly aborted server that is still present in the list, judging by last heartbeat time. The reason is that server heartbeat and server timeout are values configurable **per server**, and the current `FetchedJobsWatcher` process has no insight into how these settings may be configured on other servers. Thus, it can't make decisions whether a particular server is aborted or not.

3. `RedisStorage` transactional feature flags are now reported to the core library depending on how the user configured them. If the user specified `UseTransactions = false`, the storage reports to the core library that it does not support transactions.

4. Deduplicated `RedisStorage` constructors. The way to determine what parameter is set is questionable.

5. When a `RedisFetchedJob` is requeued, it now publishes its id to the subscription channel for immediate execution.

### CODE REVIEW

File: Hangfire.Redis.Tests/RedisFetchedJobFacts.cs [hunk 1]
```diff
@@ -90,6 +96,27 @@ public void RemoveFromQueue_RemovesOnlyJobWithTheSpecifiedId()
...
+                fetchedJob.RemoveFromQueue();
+
+                // Assert
+                Assert.Equal(1, redis.ListLength("{hangfire}:queue:my-queue:dequeued"));
+                Assert.Equal("job-id", (string)redis.ListRightPop("{hangfire}:queue:my-queue:dequeued"));
+            });
+        }
+
         [Fact, CleanRedis]
         public void RemoveFromQueue_RemovesOnlyOneJob()
```
Comment (Lexy2) [inline Hangfire.Redis.Tests/RedisFetchedJobFacts.cs:121]: I'm not really sure what this test is supposed to test. Having two jobs with the same `ID` in the `dequeued` list is probably a result of a bug. You can't save other dequeued job parameters in the job's hash. Completing the job will remove only one job from this list, but then what will happen to the other dequeued instance?

File: Hangfire.Redis.StackExchange/RedisConnection.cs [hunk 2]
```diff
@@ -150,13 +150,17 @@ public override bool GetSetContains([NotNull] string key, [NotNull] string value
...
             var storedParameters = new Dictionary<string, string>(parameters)
             {
-                { "Queue", invocationData.Queue },
                 { "Type", invocationData.Type },
                 { "Method", invocationData.Method },
                 { "ParameterTypes", invocationData.ParameterTypes },
                 { "Arguments", invocationData.Arguments },
                 { "CreatedAt", JobHelper.SerializeDateTime(createdAt) }
             };
+
```
Comment (Lexy2) [inline Hangfire.Redis.StackExchange/RedisConnection.cs:159]: I'm not sure why this change is not in the upstream. I saw that you merged that PR, but it still shows this as a change.

File: Hangfire.Redis.Tests/RedisFetchedJobFacts.cs (hunk 1 above)
Comment (marcoCasamento) [reply Hangfire.Redis.Tests/RedisFetchedJobFacts.cs:121]: Agree. I see that the test is part of the original test suite, It probably made sense at that time. There are no checks about the values in the dequeued list so it's possible that this list contains duplicate values, however, I agree, I believe it's just the result of a bug. Do you want to remove it ?

File: Hangfire.Redis.StackExchange/RedisConnection.cs [hunk 3]
```diff
@@ -150,13 +150,17 @@ public override bool GetSetContains([NotNull] string key, [NotNull] string value
...
-                { "Queue", invocationData.Queue },
                 { "Type", invocationData.Type },
                 { "Method", invocationData.Method },
                 { "ParameterTypes", invocationData.ParameterTypes },
                 { "Arguments", invocationData.Arguments },
                 { "CreatedAt", JobHelper.SerializeDateTime(createdAt) }
             };
+            
+            if (invocationData.Queue != null)
+            {
```
[the remark this answers is not in the document]
Comment (marcoCasamento) [reply Hangfire.Redis.StackExchange/RedisConnection.cs:161]: you sure it's not a local issue ? It does seems present in master https://github.com/marcoCasamento/Hangfire.Redis.StackExchange/blame/master/Hangfire.Redis.StackExchange/RedisConnection.cs

[the remark this answers is not in the document]
Comment (Lexy2) [reply Hangfire.Redis.StackExchange/RedisConnection.cs:161]: Maybe my fork is somehow not up-to-date, but it pops up in the changes, I don't understand why.

File: Hangfire.Redis.Tests/RedisFetchedJobFacts.cs (hunk 1 above)
Comment (Lexy2) [reply Hangfire.Redis.Tests/RedisFetchedJobFacts.cs:121]: Will do.

File: Hangfire.Redis.StackExchange/RedisConnection.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lexy2) [reply Hangfire.Redis.StackExchange/RedisConnection.cs:161]: Synced forks, removed duplicate code.

**Comment by Lexy2** [comment]
@marcoCasamento how's the testing going?

**Comment by Lexy2** [comment]
@marcoCasamento can I help you with any concerns related tot his PR?

**Comment by Lexy2** [comment]
Given that the release with the previous fix is unlisted, I think it's worth publishing it now and dealing with regressions when they arrive.

I am running a forked build of this release in my environment and I'm seeing no issues.

So I'd go for it, wait for issues, and fix them forward. 🤝

**Comment by Lexy2** [comment]
And when you approve & merge, please also publish an updated release to NuGet.

**Comment by FixRM** [comment]
Thank you guys!
